src/routers/alpha-router/gas-models/mixedRoute/mixed-route-heuristic-gas-model.ts

```diff
@@ -0,0 +1,271 @@
+    // of the quote token in order to produce a gas adjusted amount.
+    // We do return a gas use in USD however, so we still convert to usd.
+    const nativeCurrency = WRAPPED_NATIVE_CURRENCY[chainId]!;
+    if (token.equals(nativeCurrency)) {
+      const estimateGasCost = (
+        routeWithValidQuote: MixedRouteWithValidQuote
+      ): {
+        gasEstimate: BigNumber;
+        gasCostInToken: CurrencyAmount;
+        gasCostInUSD: CurrencyAmount;
+      } => {
+        const { totalGasCostNativeCurrency, baseGasUse } = this.estimateGas(
+          routeWithValidQuote,
+          gasPriceWei,
+          chainId
+        );
+
+        const token0 = usdPool.token0.address == nativeCurrency.address;
+
+        const nativeTokenPrice = token0
+          ? usdPool.token0Price
+          : usdPool.token1Price;
+
+        const gasCostInTermsOfUSD: CurrencyAmount = nativeTokenPrice.quote(
+          totalGasCostNativeCurrency
+        ) as CurrencyAmount;
+
+        return {
+          gasEstimate: baseGasUse,
+          gasCostInToken: totalGasCostNativeCurrency,
+          gasCostInUSD: gasCostInTermsOfUSD,
+        };
+      };
+
+      return {
+        estimateGasCost,
+      };
+    }
+
+    // If the quote token is not in the native currency, we convert the gas cost to be in terms of the quote token.
+    // We do this by getting the highest liquidity <quoteToken>/<nativeCurrency> pool. eg. <quoteToken>/ETH pool.
+    const nativeV3Pool: Pool | null = await getHighestLiquidityV3NativePool(
+      chainId,
+      token,
+      V3poolProvider
+    );
+
+    let nativeV2Pool: Pair | null;
+    if (V2poolProvider) {
+      /// MixedRoutes
+      nativeV2Pool = await getEthV2Pool(chainId, token, V2poolProvider);
+    }
+
+    const usdToken =
+      usdPool.token0.address == nativeCurrency.address
+        ? usdPool.token1
+        : usdPool.token0;
+
+    const estimateGasCost = (
+      routeWithValidQuote: MixedRouteWithValidQuote
+    ): {
+      gasEstimate: BigNumber;
+      gasCostInToken: CurrencyAmount;
+      gasCostInUSD: CurrencyAmount;
+    } => {
+      const { totalGasCostNativeCurrency, baseGasUse } = this.estimateGas(
+        routeWithValidQuote,
+        gasPriceWei,
+        chainId
+      );
+
+      if (!nativeV3Pool && !nativeV2Pool) {
+        log.info(
+          `Unable to find ${nativeCurrency.symbol} pool with the quote token, ${token.symbol} to produce gas adjusted costs. Route will not account for gas.`
+        );
+        return {
+          gasEstimate: baseGasUse,
+          gasCostInToken: CurrencyAmount.fromRawAmount(token, 0),
+          gasCostInUSD: CurrencyAmount.fromRawAmount(usdToken, 0),
+        };
+      }
+
+      /// we will use nativeV2Pool for fallback if nativeV3 does not exist
+      /// can use ! here because we return above if v3Pool and v2Pool are null
+      const nativePool =
+        !nativeV3Pool && nativeV2Pool ? nativeV2Pool : nativeV3Pool!;
+
```

Comment:
The native v3 pool could exist but have no liquidity. Think we should add some sort of heuristic too to check TVL USD is at least some number.. what number I don't know though.. maybe around 500-1000?

Reply:
We don't have access to TVLUSD here, that's only from the subgraph. We can check properties like `Pool.liquidity` or `Pair.getLiquidityValue()`, but I'm hesitant because any number we compare it to is relative depending on the token right?

Reply:
Mmm yes thats true.. Ok this is probably fine for now